Auto/IR/SMT.lean

```diff
@@ -415,29 +426,66 @@ section
   def hIn (e : ω) : TransM ω Bool := do
     return (← getH2lMap).contains e
 
-  /-- Used for e.g. bound variables -/
-  partial def disposableName : TransM ω String := do
+  def binderNamePrefixFromSort (sort : SSort) : String :=
+    match sort with
+    | SSort.bvar _ => "bvar"
+    | SSort.app (SIdent.symb s) _ => s.take 1
+    | SSort.app (SIdent.indexed s _) _ => s.take 1
+
+  /-- Used for e.g. bound variables. -/
+  partial def disposableName (sortHint : SSort): TransM ω String := do
     let l2hMap ← getL2hMap
     let idx ← getIdx
-    let currName := s!"smtd_{idx}"
+    let mut currName := s!"{binderNamePrefixFromSort sortHint}{idx}"
+    -- Try to avoid collisions with other identifiers
     if l2hMap.contains currName then
-      throwError "disposableName :: Unexpected error"
+      currName := s!"{currName}_{idx}"
+      if l2hMap.contains currName then
+        throwError "disposableName :: Unexpected error - binder disposable name already exists!"
     setIdx (idx + 1)
     return currName
 
+  def smtNameFromHint (nameHint : Option Expr) (default: String): TransM ω String := do
+    match nameHint with
+    | none => return default
+    | some nameHint => do
+      match nameHint with
+      | Expr.const name _ => return name.toString
+      | Expr.fvar fvarId =>
+        let lctx ← getLCtx
+        match lctx.find? fvarId with
+        | some localDecl => return localDecl.userName.toString
+        | none => return default
+      -- We might have different applications of the same polymorphic function;
+      -- to disambiguate, we use the argument names, not just the function names.
+      | Expr.app f args =>
+        let fname ← smtNameFromHint f default
+        let argname ← smtNameFromHint args "arg"
+        return s!"{fname}_{argname}"
+      | Expr.proj typeName idx _ => return s!"{typeName.toString}_proj_{idx}"
+      | _ =>
+        trace[auto.lamReif.printValuation] "smtNameFromHint :: Unexpected hint; saw {nameHint.ctorName} {nameHint} ({nameHint.dbgToString})"
+        return default
+
   /--
     Turn high-level construct into low-level symbol
     Note that this function is idempotent
+    `nameHint` is an expression from which we can extract a name.
   -/
-  partial def h2Symb (cstr : ω) : TransM ω String := do
+  partial def h2Symb [ToString ω] (cstr : ω) (nameHint : Option Expr := none) : TransM ω String := do
     let l2hMap ← getL2hMap
     let h2lMap ← getH2lMap
     if let .some name := h2lMap.find? cstr then
       return name
     let idx ← getIdx
-    let currName : String := s!"smti_{idx}"
-    if l2hMap.contains currName then
-      throwError "h2Symb :: Unexpected error"
+    let defaultName : String := s!"smti_{idx}"
+    let mut currName ← smtNameFromHint nameHint defaultName
+    trace[auto.lamReif.printValuation] "smti_{idx} := {currName} (from expr: {nameHint} | lam: {cstr})"
```

Comment:
``auto.lamReif.printValuation`` is inappropriate. May be register a new trace class, e.g. ``auto.smt.h2symb``